src/iop/hazeremoval.c

```diff
@@ -0,0 +1,962 @@
+  b = t;
+}
+
+// calculate the one-dimensional moving average over a window of size 2*w+1
+// input array x has stride 1, output array y has stride stride_y
+static inline void box_mean_1d(int N, const float *x, float *y, size_t stride_y, int w)
+{
+  float m = 0, n_box = 0;
+  for (int i = 0, i_end = min_i(w+1, N); i < i_end; i++)
+  {
+    m += x[i];
+    n_box++;
+  }
+  for (int i = 0; i < N; i++) {
+    y[i * stride_y] = m / n_box;
+    if (i - w >= 0)
+    {
+      m -= x[i - w];
+      n_box--;
+    }
+    if (i + w + 1 < N)
+    {
+      m += x[i + w + 1];
+      n_box++;
+    }
+  }
+}
+
+// calculate the two-dimensional moving average over a box of size (2*w+1) x (2*w+1)
+// does the calculation in-place if input and ouput images are identical
+// this function is always called from a OpenMP thread, thus no parallelization
+static void box_mean(gray_image img1, gray_image img2, int w)
+{
+  gray_image img2_bak;
+  if (img1.data == img2.data)
+  {
+    img2_bak = new_gray_image(max_i(img2.width, img2.height), 1);
+    for (int i1 = 0; i1 < img2.height; i1++)
+    {
+      memcpy(img2_bak.data, img2.data + (size_t)i1 * img2.width, sizeof(gray_pixel) * img2.width);
+      box_mean_1d(img2.width,
+                  img2_bak.data,
+                  img2.data + (size_t)i1 * img2.width, 1,
+                  w);
+    }
+  }
+  else
+  {
+    for (int i1 = 0; i1 < img1.height; i1++)
+      box_mean_1d(img1.width,
+                  img1.data + (size_t)i1 * img1.width,
+                  img2.data + (size_t)i1 * img2.width, 1,
+                  w);
+    img2_bak = new_gray_image(1, img2.height);
+  }
+  for (int i0 = 0; i0 < img1.width; i0++)
+  {
+    for (int i1 = 0; i1 < img1.height; i1++)
+      img2_bak.data[i1] = img2.data[i0 + (size_t)i1 * img2.width];
+    box_mean_1d(img1.height,
+                img2_bak.data,
+                img2.data + i0, img1.width,
+                w);
+  }
+  free_gray_image(&img2_bak);
+}
+
+// calculate the one-dimensional moving maximum over a window of size 2*w+1
+// input array x has stride 1, output array y has stride stride_y
+static inline void box_max_1d(int N, const float *x, float *y, size_t stride_y, int w)
+{
+  float m = -(INFINITY);
+  for (int i = 0, i_end = min_i(w+1, N); i < i_end; i++)
+    m = max_f(x[i], m);
+  for (int i = 0; i < N; i++) {
+    y[i * stride_y] = m;
+    if (i - w >= 0 && x[i - w] == m)
+    {
+      m = -(INFINITY);
+      for (int j = max_i(i-w+1, 0), j_end = min_i(i+w+2, N); j < j_end; j++)
+        m = max_f(x[j], m);
+    }
+    if (i + w + 1 < N)
+      m = max_f(x[i + w + 1], m);
+  }
+}
+
+// calculate the two-dimensional moving maximum over a box of size (2*w+1) x (2*w+1)
+// does the calculation in-place if input and ouput images are identical
+static void box_max(gray_image img1, gray_image img2, int w)
+{
+  gray_image img2_bak;
+  if (img1.data == img2.data)
+  {
+#ifdef _OPENMP
+#pragma omp parallel default(none) shared(img1, img2, w) private(img2_bak)
```

Comment:
if this is a hint to the reader, then please write it as a comment :)
Though this would be the only place where such comment exists, thus it will only create confusion.
So just drop `shared()` unless they are **really** needed.
If they are needed, try marking `img1` / `img2` as `const` first.